### PR TITLE
Support disabling validators.

### PR DESCRIPTION
**Pull request by dvc94ch**
Instead of forcing a new era when the staking module slashes a validator it disables the validator. The default implementation for this forces a new era, but the babe module should be able to do it's own book keeping.

### CODE REVIEW

File: srml/session/src/lib.rs [hunk 1]
```diff
@@ -343,12 +374,21 @@ mod tests {
...
 	thread_local!{
 		static NEXT_VALIDATORS: RefCell<Vec<u64>> = RefCell::new(vec![1, 2, 3]);
+		static DISABLED_VALIDATORS: RefCell<Vec<u64>> = RefCell::new(vec![]);
 	}
 
 	pub struct TestOnSessionChange;
 	impl OnSessionChange<u64> for TestOnSessionChange {
 		fn on_session_change(_elapsed: u64, _should_reward: bool) {
 			NEXT_VALIDATORS.with(|v| Session::set_validators(&*v.borrow()));
+		DISABLED_VALIDATORS.with(|v| v.borrow_mut().clear());
```
Comment (bkchr) [inline srml/session/src/lib.rs]:
```suggestion
			DISABLED_VALIDATORS.with(|v| v.borrow_mut().clear());
```

File: srml/session/src/lib.rs [hunk 2]
```diff
@@ -118,7 +118,7 @@
 use rstd::prelude::*;
 use primitives::traits::{Zero, One, Convert};
 use srml_support::{StorageValue, StorageMap, for_each_tuple, decl_module, decl_event, decl_storage};
-use srml_support::{dispatch::Result, traits::OnFreeBalanceZero};
+use srml_support::{traits::OnFreeBalanceZero};
```
Comment (bkchr) [inline srml/session/src/lib.rs]:
```suggestion
use srml_support::traits::OnFreeBalanceZero;
```

**Comment by gavofyork** [comment]
This isn't quite right. Is it only meant as the first step with several follow-up PRs yet to come?

Staking shouldn't be implementing or responding to `OnDisable` directly at all. The only signal Staking should be responding to is `SessionWillEnd`. The only modules that should be listening are Grandpa, Babe and (in Polkadot) Parachains. Session should have a configurable type in its trait to tell it when the session `is_ended`, which for now can be a simple `block_number() % SESSION_LENGTH == 0`, but will ultimately come from Babe.

**Comment by dvc94ch** [comment]
Is SessionWillEnd distinct from OnSessionChange?

**Pull request closed without merging**